### ISSUE TITLE
Map input devices to actions

### ISSUE THREAD

**Issue by ebkalderon** [amethyst/amethyst#177]
We need some way to map discrete keyboard, mouse, or gamepad input to simple action strings through flat-text configuration files, e.g. "jump" could map to either `keyboard: Space` or `gamepad: A`.

See the [example given in the wiki](https://github.com/amethyst/amethyst/wiki/Configuration-Design#input-configuration) for details.

**Comment by unpatched** [comment]
Hi all,

The general API.  This uses 'char' as a key, the value is a Box<Fn>

`fn main() {
    let mut camera = Camera::new();
    let mut bindings = HashMap::new();
    let up = 'w';
    let down = 's';
    let right = 'd';
    let left = 'a';
    let z_in = '+';
    let z_out = '-';
    Camera::add_binding_entry(&mut bindings, up,    Box::new(move |_| camera.inc_x()));
    Camera::add_binding_entry(&mut bindings, down,  Box::new(move |_| camera.dec_x()));
    Camera::add_binding_entry(&mut bindings, right, Box::new(move |_| camera.inc_y()));
    Camera::add_binding_entry(&mut bindings, left,  Box::new(move |_| camera.dec_y()));
    Camera::add_binding_entry(&mut bindings, z_in,  Box::new(move |_| camera.zoom_in()));
    Camera::add_binding_entry(&mut bindings, z_out, Box::new(move |_| camera.zoom_out()));
}
`
The "Controller" trait.

`pub trait Controller<T> {
    fn controller_loop(&mut self, data: T, bindings: HashMap<char, Box<Fn(&mut Self)>>);
    fn add_binding_entry(bindings: &mut HashMap<char, Box<FnMut(&mut Self)>>, char: char, b_fun: Box<FnMut(&mut Self)>);
}
`

Then the impl

`impl Controller<String> for Camera {
    fn controller_loop(&mut self, mut data: String, bindings: HashMap<char, Box<Fn(&mut Self)>>) {
        for res in data.drain(..1) {
            if let Some(x) = bindings.get(&res) {
                x(self)
            }
        }
    }
    fn add_binding_entry(bindings: &mut HashMap<char, Box<FnMut(&mut Self)>>, char: char, b_fun: Box<FnMut(&mut Self)>) {
        bindings.insert(char, b_fun);
    }
}
`

I think the control loop fn could use some work. It takes a stream because it's only interested in taking however many u8's you want to throw at it. I also can't remember if I actually tested this. I'm interested in helping out with amethyst.

**Comment by derekdreery** [comment]
Would it be better to map input events to an enum, and then implement what happens for the enum on the stack (would this be better for the cache?)?

**Comment by derekdreery** [comment]
Or does it not matter since this is only once per frame

**Comment by unpatched** [comment]
> Would it be better to map input events to an enum, and then implement what happens for the enum on the stack (would this be better for the cache?)?

It sounds like experiment time! Everything else uses an enum but I thought it would also be excellent to just use arbitrary types as keys, and function calls as values!  I have a lot of boring work to do today but I could probably test this out in 12 hours or so.

> Or does it not matter since this is only once per frame

I'm not a computer scientist :(.  Someone with engine building knowledge would have to answer this question.  I'm happy to re-factor however we all agree is prudent!  I needed this bit of code since I'm using Amethyst as a video editor.  Haha.

**Comment by OvermindDL1** [comment]
Be sure to be able to handle modifier's as well, so that both 'a' and 'shift+a' do different actions, but if only 's' is bound then hitting 'shift+s' would still call s.  Ditto with Ctrl, Alt, and Meta as well.  A more generic method would be allowing for any multiple-key interaction as well.

**Comment by unpatched** [comment]
The idea with this method so far is being able to use any kind of trigger for arbitrary function calls simply by changing the type of the HashMap's Key.  In my mind the I/O API should make little to no assumptions about the device the API is handling.

The example currently uses the keyboard because it was the easiest to whip up, but you are not limited to that, infact anything that Read can be implemented for is usable.  You'll have to whip up your own controller loop for each device but we can make it more ergonomic with a macro at some point.  The controller loop's iterator is actually responsible for parsing input from streams.

If we had a keytype of String you could truly define any arbitrary series of bytes as a trigger.  char was used for brevity. Haha.

**Comment by derekdreery** [comment]
or [u8]?

**Comment by unpatched** [comment]
+1 to [u8] all in favor?

**Comment by ebkalderon** [comment]
@unpatched No, I agree with @OvermindDL that we should be using strong typing (enums) for the keys here. If you look at the wiki article referenced in the first post, we essentially reference something a lot like `VirtualKeyCode`s. We organize each key binding into the type of device, using the map names `keyboard:` or `gamepad:`. With this, we can have our input bindings for different input devices be easy to validate and distinguish (e.g. what's the difference between the "x" key and the gamepad "x" button) plus we can specify modifiers if needed.

**Comment by unpatched** [comment]
It would be trivial to implement ControlHandler for Keyboard or GamePad.  I plan on implementing it for the touchscreen interface winit provides as well.  I'm more than happy to get rid of `Box<Fn>` and replace it with `VirtualKeyCode`, but I think the HashMap method offers a great degree of diversity when you consider the number of non kb+m or gamepad controllers.

**Comment by unpatched** [comment]
If you want a PoC I'll be back in a few hours! :D  Whatever is decided I pledge my Keyboard!

**Comment by OvermindDL1** [comment]
As an example, an action mapper I built, oh, probably a decade ago worked like this (everything really was string keys, but most were interned via a flyweight so comparisons were simple pointer comparisons at runtime):

Given, say, this system had two keyboards plugged in, 1 mouse, and 3 gamepads:
You could grab a keyboard via a UUID (so you always grabbed the same one each time) or via index:
```cpp
Keyboard *keyboard = system.getDevice<Keyboard>("/dev/keyboards/<uniqueID-or-integer-index>");
```
These lookups did cause a search tree to be searched so was not 'immediately' fast, but there were generally a one-time thing at start-up or a change, and you could get a listing of everything under a tree path as well.  It was the same for a mouse, gamepad, speaker, monitor, etc...  Underneath a device like a keyboard you could query for its name, UUID, keys, etc....  A monitor exposed its name, UUID, resolutions, current-resolution, etc...

However, you could create event links:
```cpp
system.createEventMapping("/dev/keyboards/0/keychange/w", "/players/0/inputs/forward");
```
And anything that subscribed to the event via:
```cpp
system.subscribeEvent("/players/0/inputs/forward", bind(&this, onForward));
```
Would get an event structure.

An event structure (which I detailed somewhere in a specs issue as I recall) was just an `std::map` (tree map, not hash-map, tree maps are faster on a small count of things, which events almost always were, but the key was an Atom64 (a compile-time interned string to an integer, so an integer), and the value was a Variant of limited types) for such a 'forward' might be (in more modern C++ instead of the older one I had to work with at the time):
```cpp
void onForwardChange(Atom64 _eventName, VariantMap &event) {
  float speed = 0.0; // default do nothing if no understandable input
  // A keyboard will always return here a 0 or 1, a gamepad would be a range from 0.0 to 1.0 when you map to a half-axis
  if(Variant *range = event.at(INPUT_VALUE)) speed = range.getFloat(1.0); // 1.0 is the default if the variant cannot become a float
  this.setForwardVelocity(speed * this.forwardSpeedPerSecond);
}
```

If you wanted to map an event that changed the event to a format better for the reciever, you could do that too:
```cpp
system.createEventMapping("/dev/keyboards/0/keychange/w", "/players/0/inputs/forward", bind(&this, mapKeyboardEventToPlayerInputEvent));
```
Would just took the event and returned a new event of the new format you want to pass on.

The 'nodes' in the system were just a 'thing' of some sort, like `"dev"` that was mounted to the root of `""` just had its accessor functions return a device based on what was accessed, so accessing "keyboards" on "dev" on "root" (or succinctly, "/dev/keyboards") returned a thing where you could enumerate more nodes over, but those returned keyboards or nothing when queried, based on the ID that was queried.

Querying over the system mapping should not be done often, usually just to grab some things or set up event mappings, then the game runs, but it was often used for game tasks as well, like the current level could register itself at "/level/current", and a certain named entity in that level could opt to register itself under its name in the level, so it would be accessible at "/level/current/blahname", which if it were, say, a switch, then the level could map from the switch's `use` event to blowing up tnt somewhere or so, like via:
```cpp
// The level would have access to its own node, so no need to query the system root,
// give it the node you want to work from as the first argument
// The level would register this mapping when, say the switch is spawned in or at level-load or so
system.createEventMapping(thisLevel, "/superswitch/events/use", "/theBigTNTPile/blowup");
// So when blowup receives any event, it cares not anything on it, then it blows up, so when the switch is
// 'used' it sends its 'use' event, a pretty generic event used by a lot of things, contained the node that
// was 'used' as well as the entity that triggered it, not much else.
```

I always intended to build a GUI for managing all the event mappings to make level design easier, but I never did.  ^.^

The entities were, as I described in specs issues, standard ECS. 

But this style handled input action mapping, game registrations, etc...  It was just a generic event system (with some fast paths for often-called many-times-per-frame things).  :-)

And yes, accessing an invalid node in the system would throw an exception (you can query for its existence first if needed), and creating an event map would do the same from the source, but the destination did not need anything there, and a listener could still register to it (or create 'another' event mapper from it and on), they were kind of distinct namespaces, but objects could still use both as expected, like the above TNT would have a component that caused the system to register a listener at the entities 'blowup' node, even if nothing was actually there.  It worked really well though and was surprisingly fast.  :-)

**Comment by OvermindDL1** [comment]
Do note, the VariantMap and Variants were used for events because they could also be handled by scripts (specifically LUA in my case).

**Comment by unpatched** [comment]
Oooh, I never knew about TreeMaps.  I can make that happen no problem.  What is wanted/needed is much clearer now I will refactor and be back in about 10.5 hours.  Thank you very much.

**Comment by OvermindDL1** [comment]
> Oooh, I never knew about TreeMaps. I can make that happen no problem. What is wanted/needed is much clearer now I will refactor and be back in about 10.5 hours. Thank you very much.

Heh, yeah `std::map` in C++ is a [red-black tree](https://en.wikipedia.org/wiki/Red%E2%80%93black_tree), great for things where the key is fast to compare (like Atom64's/integers are) that are small in count (<10 to <24 or so, depending on the C++ implementation).  The Variant was a tagged union that could be anything from integers to float to doubles to matricies to quaternions to a Node handle to a Handle of an Entity and more, it was quite useful and type safe, and you could pass a visitor to it if you wanted to handle anything on it (it was Boost's variant, though now C++ has its own variant incoming that is based on Boost's), although a boost variant is just a weaker version of a Rust enum.  :-)

However the Tree of objects that registered somehow down to the root 'system' object was immensely useful in so many ways, it might be a good thing to swipe from my old engine, it was unmatched in usefulness and its (not that) slow parts only mattered rarely and at program start-up, but it was still fast enough regardless in every case I've had (it is entirely a convenience thing, which a key action map is too anyway).

An example of how it worked from my example above:
```cpp
Keyboard *keyboard = system.getNode<Keyboard>("/dev/keyboards/<uniqueID-or-integer-index>");
```
See comments in-line:
```cpp
// The 'system' variable is just a Node object, it is the 'root' in my usage
// If no template argument was given it gave it back as the root Node class
// The prefix '/' is only used when looking up multiple paths at once
Node *devNode = system.getNode("dev");
// 'devNode' is a Node just like system, although internally 'devNode' in this case is actually subclassed to
// be the DevNode class that handles the hardware via a unified interface

// You could then get the keyboards node from the dev node directly like this too
Node *keyboardsNode = devNode.getNode("keyboards");
// And a specific keyboard
Node *keyboard0Node = keyboardsNode("0");
// The numbered keyboards/mice/gamepads/etc.. could be random from run to run, depending on how
// the underlying OS enumerated them to here, UUID's were better for storing configs with, or you could
// access the "all" keyboard to register an event on all of them, say when 'a' was pressed or so

// Then you could query information about the keyboard:
String &name = keyboard0Node.getVariant("name").asString(); // if not convertable to a string it throws an exception

// The keyboards all gave back nodes if queried at:
//   "keychange": On any state change of the given key
//   "keypress": On a press and release combo, fired before "release"
//   "press": Fired on a press down
//   "release": Fired on a release up
// And a few other helper ones.  You could also listen to an event on, say, 'keychange' directly to subscribe
// to every-single event from any key change on the keyboard. which, say, a textfield would do when it
// got focus, and unsubscribe when it was blurred/lost-focus.

// These two are equivalent, though if registering a lot of keys then holding the keychange node and
// registering on it directly is more efficient.
keyboard0Node.subscribeEvent("keychange/a", listener); // Remember, a '/' in the path means before it must be a node or the call throws an exception
Node *keyboard0KeyChangeNode = keyboard0Node.getNode("keychange");
keyboard0KeyChangeNode.subscribeEvent("a", listener);
```

**Comment by OvermindDL1** [comment]
Oh, the general interface to a Node was something like:
```c++
class Node {

  // get a variant at a path, `getNode` called this to get the variant, then checked if the variant was a node
  // handle, if not a node handle it throws, if a node handle then it checks if a valid/active node, if not it
  // throws, if so it returns it.
  // `getVariant` got the first nodename in the path and looked up in some information to get a variant
  // with that name, for 'keyboards' for example it would query the KeyboardController system for a
  // keyboard with the given index id or UUID or so and return it if it existed, else it would return a Variant
  // of Nothing.  If more path then it would then call 'getVariant' on the keyboard that was returned, which
  // it would continue down its path as expected.
  // `getVariant` is the base method in the node tree.
  Variant getVariant(string &path);

  Node &getNode(string &path); // gets a path via getVariant and throw or returns if a valid Node or not

  // Few other minor helpers, but 'getVariant' is the only real important one that everything else could be
  // built up from.  Specific nodes like for a Keyboard or a Level would have their own functions too of
  // course if they could be cast safely first, but everything 'should' be accessible via the getVariant call, if
  // not always the most efficiently unlike direct calls, but should always be accessible that way, it was
  // especially important for scripting systems to be able to access anything that way.
```

**Comment by unpatched** [comment]
I will implement this as the consensus seems to have been drawn to go with ODL1's well thought out option, I have no horse in this race and am just happy to be able to contribute. Haha.

**Comment by ebkalderon** [comment]
@unpatched Thanks, we are glad to have you here! Any help is greatly appreciated.

**Comment by unpatched** [comment]
std::collections::TreeMap is "experimental", are there any objections to still using it?

**Comment by unpatched** [comment]
Ok so ~ 12 hours late but here we are!

```
extern crate amethyst;

use std::collections::HashMap;
use amethyst::VirtualKeyCode;

pub struct Bindings<T>(HashMap<Trigger<T>, Box<Fn()>>);

impl Bindings<VirtualKeyCode> {
    pub fn new() -> Bindings<VirtualKeyCode> {
        Bindings(HashMap::new())
    }
    pub fn test() {
        println!("It works!");
    }
}

pub trait ControlHandler<T> {
    fn controller_loop(&mut self,
                       data: Trigger<T>,
                       bindings: &HashMap<Trigger<T>, Box<Fn(&mut Self)>>);
    fn add_binding_entry(&mut self,
                         bindings: &mut HashMap<Trigger<T>, Box<Fn(&mut Self)>>,
                         trigger: Trigger<T>,
                         b_fun: Box<Fn(&mut Self)>);
}

impl ControlHandler<VirtualKeyCode> for Bindings<VirtualKeyCode> {
    fn controller_loop(&mut self,
                       data: Trigger<VirtualKeyCode>,
                       bindings: &HashMap<Trigger<VirtualKeyCode>, Box<Fn(&mut Self)>>) {
        if let Some(x) = bindings.get(&data) {
            x(self)
        }
    }
    fn add_binding_entry(&mut self,
                         bindings: &mut HashMap<Trigger<VirtualKeyCode>, Box<Fn(&mut Self)>>,
                         trigger: Trigger<VirtualKeyCode>,
                         b_fun: Box<Fn(&mut Self)>) {
        bindings.insert(trigger, b_fun);
    }
}

use std::io;
#[derive(PartialEq, Eq, Hash)]
pub struct Trigger<T>(Vec<T>);

impl Trigger<VirtualKeyCode> {
    fn new(trigger: &[VirtualKeyCode]) -> Trigger<VirtualKeyCode> {
        Trigger(trigger.to_vec())
    }
}

fn main() {
    //  Create a new bindings handler!
    let mut bh = Bindings::new();
    //  Create a HashMap to hold your mapped inputs!
    let mut bindings = HashMap::new();
    //  Create your triggers!
    let t1 = Trigger::new(&[VirtualKeyCode::A]);
    let t2 = Trigger::new(&[VirtualKeyCode::A, VirtualKeyCode::B]);
    //  Ask the binding handler to add your trigger to the map!
    bh.add_binding_entry(&mut bindings, t1, Box::new(move |_| Bindings::test()));
    bh.add_binding_entry(&mut bindings, t2, Box::new(move |_| println!("Haha")));

    //  This is the type signature of the input queue.
    let mut input_queue = Vec::<Trigger<VirtualKeyCode>>::new();

    //  I havent figured out Keyboard Electrons -> VirtualKeyCode (yet) so we're going to cheat
    //  and for this test pretend we recieved these inputs.
    let i1 = Trigger([VirtualKeyCode::A].to_vec());
    let i2 = Trigger([VirtualKeyCode::A, VirtualKeyCode::B].to_vec());

    //  Our cat is sleeping on our keyboard, let's see if the kitty's input triggers our function!
    let i3 = Trigger([VirtualKeyCode::S,
                      VirtualKeyCode::H,
                      VirtualKeyCode::A,
                      VirtualKeyCode::D,
                      VirtualKeyCode::I,
                      VirtualKeyCode::L,
                      VirtualKeyCode::A,
                      VirtualKeyCode::Y]
        .to_vec());
    //  Push our inputs into the queue.
    input_queue.push(i1);
    input_queue.push(i2);
    input_queue.push(i3);
    //  ???????????
    for vka in input_queue {
        //      Profit!
        bh.controller_loop(vka, &bindings);
    }
}
```

I do not expect that this will be the final form of what you want, but I would appreciate more input.  I know the map is not the type we want, it can still be changed.  It allows for multikey bindings.  I will implement Mouse tomorrow.  You should be able to w+m1 at least imo.

**Comment by unpatched** [comment]
TODO:  
Enumerate available devices and assign default bindings for those devices.
Implement/find YAML parser to tie in with config.

**Comment by unpatched** [comment]
A current, problem I have is if I need glium::Event or amethyst::Event to implement PartialEq, Eq, and Hash.  Not a deal breaker, but it will be required.

**Comment by unpatched** [comment]
Ok after a long sleep I figured that what you all want is actually to check out /dev/input/$Device, figure out what kind of device it is, and read events from them.  Is this correct?  (As opposed to a new mapping setup, the eventloop matching on enum is pretty optimal.

**Comment by msiglreith** [comment]
I personally expected some higher-level interface for the current `InputHandler`(https://github.com/amethyst/amethyst/blob/develop/src/ecs/resources/input.rs) and `WindowEvents`.

Instead of querying keys by checking for the individual device events
e.g `input_handler.is_key_down(VirtualKeyCode::A) || input_handler.is_mouse_button_down(Button1)`
we can define a input mapping which will add an higher-level API allowing us to do calls like
`input_handler.is_action_down(Forward)`.
Additionally, new events could be spawned by the different device events: `KeyboardPress` -> `ForwardPress`.

**Comment by derekdreery** [comment]
Ok - but that would require deciding onwhat types of input are possible. There are definitely
 - on/off
 - on since last check/off since last check
 - amout change since last check (for e.g. mouse)
 - clicks since last change
 - etc....

We need a difinitive list of these possibilities and also which input methods match which state types (e.g. keys => is_key_down, key_downs_since_last_check ...., mouse => mouse_position, mouse_move since last check ..., gamepad => button_pressed(intensity), touch (???).....)

**Comment by msiglreith** [comment]
Good point, we could maybe differ between 'discrete' and 'continuous' input: The former are key presses, mouse buttons, gamepad button, etc. Continuous in the this context then would be joystick axis and mouse positions.
`on/off` as  you mentioned would map to a discrete input action, represented as bool or integer.
`mouse position` could be split into x/y axis and represented by float or also integer, depending on the output from other input devices like joysticks.

This might be a good starting point for the API. Maybe someone has a better suggestion? I didn't think about this topic very much so far and there might be better APIs already existing which deal with this issue.

**Comment by derekdreery** [comment]
Gamepad buttons have continuous data (intensity).

**Comment by ebkalderon** [comment]
This idea of distinguishing between discrete and continuous input sounds a bit like functional reactive programming! @derekdreery, I'm not quite sure that all gamepad buttons have intensities. Some definitely are continuous, but others are discrete. For example, (using Xbox conventions) the A/B/X/Y buttons, the L/R shoulder buttons, the Back button, and the Start button don't measure continuous input AFAIK.

**Comment by derekdreery** [comment]
@ebkalderon actually thinking about this at this point it doesn't matter about the minutiae of how controllers work, so ignore my previous comment :).

**Comment by Xaeroxe** [comment]
So uh I kind of already implemented something like "Actions" in #247.  I see a lot of desires expressed here and different ways to design the API.  I think the API I designed covers the use cases given here, although maybe not exactly as the writers here thought it would.  I'd be curious to get feedback on my implementation from those who wrote here, as I'm going to be doing a phase 2 to enhance this API and add the bindings to the asset system (see this [comment](https://github.com/amethyst/amethyst/issues/245#issuecomment-309470760)).  My API is much simpler than the proposed ones but personally I think that's one of its strengths.

**Comment by Xaeroxe** [comment]
26 days have passed and no further comments have been made, I'm closing this issue for the time being.
